Structure_Engine/Convert/PanelToFEMesh.cs

```diff
@@ -39,58 +38,54 @@ public static partial class Convert
+        [Input("tolerance", "Tolerance used to cull duplicates from the control points of the Panel outline.")]
+        [Output("feMesh", "FEMesh converted from a Panel.")]
+        [PreviousVersion("7.3", "BH.Engine.Structure.Convert.PanelToFEMesh(BH.oM.Structure.Elements.Panel)")]
+        public static FEMesh PanelToFEMesh(this Panel panel, double tolerance = Tolerance.MacroDistance)
         {
+            // Null and invalid checks
             if (panel.IsNull())
             {
                 return null;
             }
-            List<Point> points = new List<Point>();
-            List<Edge> edges = panel.ExternalEdges;
-            List<Face> faces = new List<Face>();
-            if (!panel.IsPlanar(true, Tolerance.MacroDistance))
+            if (!panel.IsPlanar(true, tolerance))
             {
                 Base.Compute.RecordError("Panel is not planar and therefore cannot be converted to an FEMesh.");
                 return null;
             }
             if (panel.Openings.Count > 0)
             {
-                Base.Compute.RecordError("This method does not support Panels with Openings");
+                Base.Compute.RecordError("This method does not support Panels with Openings.");
                 return null;
             }
-            if (edges.Count > 4)
-            {
-                Base.Compute.RecordError("Panel contains more than 4 Edges");
-                return null;
-            }
-            foreach (Edge edge in edges)
-            {
-                ICurve curve = edge.Curve;
-                points.AddRange(Geometry.Convert.IToPolyline(curve).ControlPoints);
-            }
-            int count = points.Distinct().Count();
+
             Face face = new Face();
-            if (count > 4)
-            {
-                Base.Compute.RecordError("Panel contains more than four control points.");
-                return null;
-            }
-            if (count == 4)
+            if (panel.IsOutlineQuad())
             {
                 face = Geometry.Create.Face(0, 1, 2, 3);
             }
-            else if (count == 3)
+            else if (panel.IsOutlineTriangular())
             {
                 face = Geometry.Create.Face(0, 1, 2);
             }
-            faces.Add(face);
-            Mesh mesh = Geometry.Create.Mesh(points.Distinct(), faces);
+            else
+            {
+                Base.Compute.RecordError("Panel is not a planar quadilateral or triangular.");
+                return null;
+            }
+
+            PolyCurve outline = panel.OutlineCurve().Curves.IJoin()[0];
+            List<Point> points = outline.DiscontinuityPoints();
+            points = points.CullDuplicates(tolerance).ISortAlongCurve(outline);
+
+            List<Face> faces = new List<Face>() { face };
+
+            Mesh mesh = Geometry.Create.Mesh(points, faces);
             FEMesh feMesh = new FEMesh();
             feMesh = Create.FEMesh(mesh, null, null, panel.Name);
+
```

Comment:
Could just provide the panel property to the constructor here, if it is null or not. After all, it is null by default, so only setting it if it is not null makes no difference.

Reply:
Added.